minecraft-common/src/main/java/com/sigmundgranaas/forgero/minecraft/common/toolhandler/ToolBlockHandler.java

```diff
@@ -6,72 +6,80 @@
-            var handler = BlockHandlerCache.computeIfAbsent(key, () -> createHandler(container, world, pos, player));
-            return Optional.of(handler);
-        }
-        return Optional.empty();
-    }
-
-    public static ToolBlockHandler createHandler(PropertyContainer container, BlockView world, BlockPos pos, PlayerEntity player) {
-        var blockMiningData = container.stream().features().filter(data -> data.type().equals(BLOCK_BREAKING_PATTERN_KEY) || data.type().equals(VEIN_MINING_KEY)).toList();
-        BlockBreakingHandler handler;
-        var data = blockMiningData.get(0);
-        if (data.type().equals(BLOCK_BREAKING_PATTERN_KEY)) {
-            handler = new BlockBreakingHandler(new PatternBreakingStrategy(data));
-        } else {
-            handler = new BlockBreakingHandler(new VeinMiningStrategy(data));
-        }
-        return new ToolBlockHandler(handler.getAvailableBlocks(world, pos, player), pos, handler.getHardness(world.getBlockState(pos), pos, world, player));
-    }
-
-    public float getHardness() {
-        return hardness;
-    }
-
-    public void handle(Consumer<BlockInfo> consumer) {
-        availableBLocks.forEach(consumer);
-    }
-
-    public void handleExceptOrigin(Consumer<BlockInfo> consumer) {
-        availableBLocks.stream().filter(info -> !info.pos.equals(originPos)).forEach(consumer);
-    }
-
-    public record BlockInfo(BlockPos pos, BlockState state) implements CacheAbleKey {
-        @Override
-        public String key() {
-            return pos.asLong() + state.toString();
-        }
-    }
+	public static ToolBlockHandler EMPTY = new ToolBlockHandler(new TreeMap<>(), new BlockPos(0, 0, 0), 0f);
+	public static String BLOCK_BREAKING_PATTERN_KEY = "BLOCK_BREAKING_PATTERN";
+	public static String VEIN_MINING_KEY = "VEIN_MINING";
+
+	private final List<BlockInfo> availableBlocks;
+	private final BlockPos originPos;
+	private final float hardness;
+
+	public ToolBlockHandler(SortedMap<BlockPos, BlockState> availableBlocks, BlockPos originPos, float hardness) {
+		this.availableBlocks = new ArrayList<>();
+		availableBlocks.forEach((blockPos, blockState) -> {
+			this.availableBlocks.add(new BlockInfo(blockPos, blockState));
+		});
+
+		this.originPos = originPos;
+		this.hardness = hardness;
+	}
+
+	public static Optional<ToolBlockHandler> of(PropertyContainer container, BlockView world, BlockPos pos, PlayerEntity player) {
+		boolean has = ContainsFeatureCache.check(PropertyTargetCacheKey.of(container, BLOCK_BREAKING_PATTERN_KEY)) || ContainsFeatureCache.check(PropertyTargetCacheKey.of(container, VEIN_MINING_KEY));
+		if (has) {
+			var key = new BlockHandlerCache.BlockStateCacheKey(new BlockInfo(pos, world.getBlockState(pos)), container, Direction.getEntityFacingOrder(player));
+			var handler = BlockHandlerCache.computeIfAbsent(key, () -> createHandler(container, world, pos, player));
```

Comment:
Gonna comment here although this was an issue before your rewrite, the cache either needs to remove the exact entry used for this calculation after it is handled or just cleared in general. Using the key should be sufficient to remove the entry.

If you place a block right where the last one was, the whole vein mining list will be retrieved. I can also fix this issue myself, so feel free to ignore.

Reply:
I can't really figure out how you want me to fix this. Maybe I'm just too tired at the moment.
Not too familiar with this class, so not sure what needs to happen.

Reply:
As discussed in the Forgero Discord server, this will be moved to a separate issue/PR.
https://discord.com/channels/981828752029925408/981828752659083267/1078377488062890084